Remotion/Data/DomainObjects/Mapping/Builder/ClassDefinitionBuilderNode.cs

```diff
@@ -0,0 +1,111 @@
+  {
+    public ClassDefinitionBuilderNode? BaseClass { get; }
+
+    private readonly List<ClassDefinitionBuilderNode> _childClasses = new();
+
+    private ClassDefinition? _classDefinition;
+    private bool _isBuilt;
+    private bool _isReadOnly;
+
+    public ClassDefinitionBuilderNode (Type type, ClassDefinitionBuilderNode? baseClass)
+        : base(type)
+    {
+      if (!type.IsClass)
+        throw new InvalidOperationException("The specified type must be a class.");
+
+      BaseClass = baseClass;
+      BaseClass?.AddChildClass(this);
+    }
+
+    public IReadOnlyList<ClassDefinitionBuilderNode> ChildClasses => _childClasses;
+
+    public ClassDefinition? ClassDefinition => _classDefinition;
+
+    /// <inheritdoc />
+    public override bool IsLeafNode => ChildClasses.Count == 0;
+
+    /// <inheritdoc />
+    public override void SetReadOnly ()
+    {
+      if (_isReadOnly)
+        return;
+
+      // Set read-only right away to prevent loops in the following recursion
+      _isReadOnly = true;
+
+      BaseClass?.SetReadOnly();
+      foreach (var childClass in ChildClasses)
+        childClass.SetReadOnly();
+
+      if (_classDefinition != null)
+        _classDefinition.SetDerivedClasses(ChildClasses.Select(e => e.ClassDefinition).Where(e => e is { BaseClass: { } })!);
+    }
+
+    /// <inheritdoc />
+    public override void BuildTypeDefinition (IMappingObjectFactory mappingObjectFactory)
+    {
+      ArgumentUtility.CheckNotNull("mappingObjectFactory", mappingObjectFactory);
+      if (_isReadOnly)
+        throw new InvalidOperationException("Cannot build type definition as the builder node is read-only.");
+      if (_isBuilt)
+        return;
+
+      ClassDefinition? baseClass = null;
+      if (BaseClass != null && !ReflectionUtility.IsInheritanceRoot(Type))
+      {
+        BaseClass.BuildTypeDefinition(mappingObjectFactory);
+        baseClass = BaseClass.ClassDefinition;
+      }
+
+      _classDefinition = mappingObjectFactory.CreateClassDefinition(Type, baseClass);
+      _isBuilt = true;
+    }
+
+    /// <inheritdoc />
+    public override TypeDefinition? GetBuiltTypeDefinition ()
+    {
+      if (!_isReadOnly)
+        throw new InvalidOperationException("Cannot get the built type definition from a non read-only builder node.");
+
+      return _classDefinition;
+    }
+
+    public void AddChildClass (ClassDefinitionBuilderNode childClass)
+    {
+      ArgumentUtility.CheckNotNull("childClass", childClass);
+      if (_isReadOnly)
+        throw new InvalidOperationException("Cannot add a child class as the builder node is read-only.");
+      if (childClass.BaseClass != this)
+        throw new InvalidOperationException("Cannot add the specified child class as it has an incorrect base class.");
+      if (_childClasses.Contains(childClass))
+        throw new InvalidOperationException("Cannot add the specified child class as it is already a child.");
```

Comment:
In the second case `if (_childClasses.Contains(childClass)) throw ...;` the argument is not the problem but rather that it has already been set. I would use `InvalidOperationException` in this case instead of `ArgumentException`. What do you think?

Reply:
Read: "Here I'm not sure"
:D 

The thing, I would try find precedence in the code base. hmm... 
https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.add?view=net-6.0
Dictionary does ArgumentException for duplicate key. I think this is close enough to the semantic of the API, superseeding the fact that the callsite "misused" the API. Unless we define the collection-aspect of Node as completely subservant. If the Children where a collection property with uniqueness semantics, the ArgumentException would be the correct choice without discussion. Given the new insights, I tend to go with my initial gut reaction towards ArgEx. Is this sound reasoning?